shell/browser/extensions/api/cryptotoken_private/cryptotoken_private_api.cc

```diff
@@ -0,0 +1,306 @@
+      cryptotoken_private::CanAppIdGetAttestation::Params::Create(*args_);
+  EXTENSION_FUNCTION_VALIDATE(params);
+
+  const GURL origin_url(params->options.origin);
+  if (!origin_url.is_valid()) {
+    return RespondNow(Error(extensions::ErrorUtils::FormatErrorMessage(
+        "Security origin * is not a valid URL", params->options.origin)));
+  }
+  const url::Origin origin(url::Origin::Create(origin_url));
+
+  const std::string& app_id = params->options.app_id;
+
+  // If the appId is permitted by the enterprise policy then no permission
+  // prompt is shown.
+  // Profile* const profile = Profile::FromBrowserContext(browser_context());
+  // const PrefService* const prefs = profile->GetPrefs();
+  // const base::ListValue* const permit_attestation =
+  //     prefs->GetList(prefs::kSecurityKeyPermitAttestation);
+
+  // for (const auto& entry : permit_attestation->GetList()) {
+  //   if (entry.GetString() == app_id)
+  //     return RespondNow(OneArgument(base::Value(true)));
+  // }
+
+  // If the origin is blocked, reject attestation.
+  if (device::fido_filter::Evaluate(
+          device::fido_filter::Operation::MAKE_CREDENTIAL, origin.Serialize(),
+          /*device=*/absl::nullopt, /*id=*/absl::nullopt) ==
+      device::fido_filter::Action::NO_ATTESTATION) {
+    return RespondNow(OneArgument(base::Value(false)));
+  }
+
+  // If prompting is disabled, allow attestation because that is the historical
+  // behavior.
+  if (!base::FeatureList::IsEnabled(
+          ::features::kSecurityKeyAttestationPrompt)) {
+    return RespondNow(OneArgument(base::Value(true)));
+  }
+
+#if defined(OS_WIN)
+  // If the request was handled by the Windows WebAuthn API on a version of
+  // Windows that shows an attestation permission prompt, don't show another
+  // one.
+  //
+  // Note that this does not account for the possibility of the
+  // WinWebAuthnApi having been disabled by a FidoDiscoveryFactory override,
+  // which may be done in tests or via the Virtual Authenticator WebDriver
+  // API.
+  if (base::FeatureList::IsEnabled(device::kWebAuthUseNativeWinApi) &&
+      device::WinWebAuthnApi::GetDefault()->IsAvailable() &&
+      device::WinWebAuthnApi::GetDefault()->Version() >=
+          WEBAUTHN_API_VERSION_2) {
+    return RespondNow(OneArgument(base::Value(true)));
+  }
+#endif  // defined(OS_WIN)
+
+  // Otherwise, show a permission prompt and pass the user's decision back.
+  const GURL app_id_url(app_id);
+  EXTENSION_FUNCTION_VALIDATE(app_id_url.is_valid());
+
+  auto* contents = electron::api::WebContents::FromID(params->options.tab_id);
+  if (!contents || !contents->web_contents()) {
+    return RespondNow(Error("cannot find specified tab"));
+  }
+
+  // permissions::PermissionRequestManager* permission_request_manager =
+  //     permissions::PermissionRequestManager::FromWebContents(web_contents);
+  //     nullptr;
+  // if (!permission_request_manager) {
+  return RespondNow(Error("no PermissionRequestManager"));
```

Comment:
Hm... it looks like we always respond with an error here. Do we need to expose something through our permission check handlers?

Reply:
This API isn't actually used by the U2F extension, I think it might be a private API that some google.com sites use 🤔  So I wasn't too concerned with wiring this up to anything in particular.

Reply:
For other extensions APIs that I've copied from Chrome, I've stripped out the bits we explicitly don't want to support. If we don't care about this API at all, maybe replace the entire implementation with just `return RespondNow(Error("API not supported in Electron"))` or so?

Reply:
I did that but in order to make diffing easier in the future I left the old implementation here with `#if 0` wrappers

Reply:
hm ok, i'm not sure that makes diffing easier but sure.